### PR TITLE
Feature/issue 651- Remote Control - Seat

### LINKED ISSUE #651: [SDL 0105] Remote Control - Seat
Proposal: [Remote Control - Seat](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0105-remote-control-seat.md)

> This proposal allows mobile applications to control powered seats within a vehicle (while the vehicle is in parking). A mobile application can send a remote control command to adjust seat positions, turn on or off cooled/heated seat, or change massage settings. The application can also get the status of each seat as well as receive notifications once the status of any seat changes.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/314

### PR DESCRIPTION
**Pull request by mrapitis**
Fixes #651 

This PR is **ready** for review.

## Risk

This PR makes **minor** API changes.

## Testing Plan

Added new unit tests and expanded existing ones

## Summary

Added RPC's, structs, enums as defined in the Remote Control - Seat proposal.

## CLA

- [x] I have signed the CLA

### CODE REVIEW

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageModeData.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+public class MassageModeData extends RPCStruct{
+    public static final String KEY_MASSAGE_ZONE = "massageZone";
+    public static final String KEY_MASSAGE_MODE = "massageMode";
+
+    /**
+     * Constructs a new MassageModeData object
+     */
+    public MassageModeData() { }
```
Comment (bilal-alsharifi) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageModeData.java]: We started to follow a best practice to add a constructor that has all mandatory params (issue 154). Also, we started adding @nonnull annotation before all mandatory params in that constructor.

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageModeData.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    public MassageMode getMassageMode() {
+        return (MassageMode) getObject(MassageMode.class, KEY_MASSAGE_MODE);
+    }
+
+    /**
+     * Sets the massageMode portion of the MassageModeData class
+     *
+     * @param massageMode
+     */
+    public void setMassageMode(MassageMode massageMode) {
```
Comment (bilal-alsharifi) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageModeData.java]: @Nonnull should be added before mandatory params

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageModeData.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    public MassageModeData(Hashtable<String, Object> hash) {
+        super(hash);
+    }
+
+    /**
+     * Sets the massageZone portion of the MassageModeData class
+     *
+     * @param massageZone
+     */
+    public void setMassageZone(MassageZone massageZone) {
```
Comment (bilal-alsharifi) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageModeData.java]: @Nonnull should be added before mandatory params

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageCushionFirmness.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+     * <p>Constructs a new MassageCushionFirmness object indicated by the Hashtable parameter
+     * </p>
+     *
+     * @param hash
+     *            The Hashtable to use
+     */
+    public MassageCushionFirmness(Hashtable<String, Object> hash) {
+        super(hash);
+    }
+
```
Comment (bilal-alsharifi) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageCushionFirmness.java]: a constructor with mandatory params should be added

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageCushionFirmness.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    public MassageCushion getCushion() {
+        return (MassageCushion) getObject(MassageCushion.class, KEY_CUSHION);
+    }
+
+    /**
+     * Sets the firmness portion of the MassageCushionFirmness class
+     *
+     * @param firmness
+     */
+    public void setFirmness(Integer firmness) {
```
Comment (bilal-alsharifi) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageCushionFirmness.java]: @Nonnull should be added before mandatory params

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageCushionFirmness.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    public MassageCushionFirmness(Hashtable<String, Object> hash) {
+        super(hash);
+    }
+
+    /**
+     * Sets the cushion portion of the MassageCushionFirmness class
+     *
+     * @param cushion
+     */
+    public void setCushion(MassageCushion cushion) {
```
Comment (bilal-alsharifi) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageCushionFirmness.java]: @Nonnull should be added before mandatory params

File: sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/enums/MassageCushionTests.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	/**
+	 * Verifies that a null assignment is invalid.
+	 */
+	public void testNullEnum () {
+		String example = null;
+		try {
+			MassageCushion temp = MassageCushion.valueForString(example);
+            assertNull("Result of valueForString should be null.", temp);
```
Comment (bilal-alsharifi) [inline sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/enums/MassageCushionTests.java]: There is a minor spacing issue

File: sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/enums/MassageModeTests.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+		assertNotNull("OFF returned null", enumOff);
+		assertNotNull("LOW returned null", enumLow);
+		assertNotNull("HIGH returned null", enumHigh);
+	}
+
+	/**
+	 * Verifies that an invalid assignment is null.
+	 */
+	public void testInvalidEnum () {
```
Comment (bilal-alsharifi) [inline sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/enums/MassageModeTests.java]: there are some spacing issues in this block of code

File: sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/enums/MassageModeTests.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+		}
+		catch (IllegalArgumentException exception) {
+            fail("Invalid enum throws IllegalArgumentException.");
+		}
+	}
+
+	/**
+	 * Verifies that a null assignment is invalid.
+	 */
+	public void testNullEnum () {
```
Comment (bilal-alsharifi) [inline sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/enums/MassageModeTests.java]: some spacing issues here as well

File: sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/enums/MassageModeTests.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+	public void testListEnum() {
+ 		List<MassageMode> enumValueList = Arrays.asList(MassageMode.values());
+
+		List<MassageMode> enumTestList = new ArrayList<MassageMode>();
+		enumTestList.add(MassageMode.OFF);
+		enumTestList.add(MassageMode.LOW);
+		enumTestList.add(MassageMode.HIGH);
+
+		assertTrue("Enum value list does not match enum class list", 
+				enumValueList.containsAll(enumTestList) && enumTestList.containsAll(enumValueList));
```
Comment (bilal-alsharifi) [inline sdl_android/src/androidTest/java/com/smartdevicelink/test/rpc/enums/MassageModeTests.java]: also a minor spacing issue here

**Comment by bilal-alsharifi** [comment]
@mrapitis PR looks good. Though, there are some best practices that we started to follow recently for RPCs:
- We started adding a constructor that has all mandatory params as described in issue 154.
- We started adding @Nonnull annotation for all mandatory params as described in proposal 0166

**Comment by bilal-alsharifi** [comment]
Also there are some minor spacing issues here and there.

**Comment by bilal-alsharifi** [comment]
@mrapitis looks very good. Thanks.

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageCushionFirmness.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+	 * </p>
+	 *
+	 * @param hash
+	 *            The Hashtable to use
+	 */
+	public MassageCushionFirmness(Hashtable<String, Object> hash) {
+        super(hash);
+    }
+
+	public MassageCushionFirmness(@NonNull MassageCushion cushion, @NonNull Integer firmness) {
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageCushionFirmness.java]: Can you please add JavaDoc to this public method

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageModeData.java [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+	 * </p>
+	 *
+	 * @param hash
+	 *            The Hashtable to use
+	 */
+	public MassageModeData(Hashtable<String, Object> hash) {
+        super(hash);
+    }
+
+	public MassageModeData(@NonNull MassageZone massageZone, @NonNull MassageMode massageMode) {
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/MassageModeData.java]: Can you please add JavaDoc to this public method

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatControlCapabilities.java [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+	 * </p>
+	 *
+	 * @param hash
+	 *            The Hashtable to use
+	 */
+	public SeatControlCapabilities(Hashtable<String, Object> hash) {
+        super(hash);
+    }
+
+	public SeatControlCapabilities(@NonNull String moduleName) {
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatControlCapabilities.java]: Can you please add JavaDoc to this public method

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatControlCapabilities.java [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+		setValue(KEY_COOLING_LEVEL_AVAILABLE, coolingLevelAvailable);
+	}
+
+	/**
+	 * Gets the coolingLevelAvailable portion of the SeatControlCapabilities class
+	 *
+	 * @return Boolean.
+	 */
+	public Boolean getCoolingLevelAvailable() {
+        return getBoolean(KEY_COOLING_LEVEL_AVAILABLE);
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatControlCapabilities.java]: Odd spacing here

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatControlCapabilities.java [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+		setValue(KEY_MEMORY_AVAILABLE, memoryAvailable);
+	}
+
+	/**
+	 * Gets the memoryAvailable portion of the SeatControlCapabilities class
+	 *
+	 * @return Boolean.
+	 */
+	public Boolean getMemoryAvailable() {
+        return getBoolean(KEY_MEMORY_AVAILABLE);
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatControlCapabilities.java]: Odd spacing here

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatControlData.java [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+	 * </p>
+	 *
+	 * @param hash
+	 *            The Hashtable to use
+	 */
+	public SeatControlData(Hashtable<String, Object> hash) {
+        super(hash);
+    }
+
+	public SeatControlData(@NonNull SupportedSeat id) {
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatControlData.java]: Can you please add JavaDoc to this public method

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatControlData.java [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+import com.smartdevicelink.proxy.rpc.enums.SupportedSeat;
+import android.support.annotation.NonNull;
+import java.util.Hashtable;
+import java.util.List;
+
+/**
+ * Seat control data corresponds to "SEAT" ModuleType.
+ */
+
+public class SeatControlData extends RPCStruct{
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatControlData.java]: Class is riddled with odd spacing inside method lines. Please address.

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatMemoryAction.java [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+	 * </p>
+	 *
+	 * @param hash
+	 *            The Hashtable to use
+	 */
+	public SeatMemoryAction(Hashtable<String, Object> hash) {
+        super(hash);
+    }
+
+	public SeatMemoryAction(@NonNull Integer id, @NonNull SeatMemoryActionType action) {
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatMemoryAction.java]: Can you please add JavaDoc to this public method

File: sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatMemoryAction.java [hunk 19]
```diff
@@ -0,0 +1,88 @@
+package com.smartdevicelink.proxy.rpc;
+
+import com.smartdevicelink.proxy.RPCStruct;
+import com.smartdevicelink.proxy.rpc.enums.SeatMemoryActionType;
+import android.support.annotation.NonNull;
+import java.util.Hashtable;
+
+public class SeatMemoryAction extends RPCStruct{
```
Comment (joeygrover) [inline sdl_android/src/main/java/com/smartdevicelink/proxy/rpc/SeatMemoryAction.java]: Class also has spacing issues of the same type. Please address.

**Comment by bilal-alsharifi** [comment]
@mrapitis There are also still some extra spaces/tabs and some extra lines in different files in the PR.

**Comment by mrapitis** [comment]
@bilal-alsharifi @joeygrover we have updated spacing and added more documentation.  Please re-review when time permits.  Thanks.

**Pull request merged**